cedar-lean/Cedar/Partial/Response.lean

```diff
@@ -0,0 +1,171 @@
+  residuals : List Residual
+  /--
+    The `Partial.Request` that was used to compute this `Partial.Response`
+  -/
+  req : Partial.Request
+  /--
+    The `Partial.Entities` that was used to compute this `Partial.Response`
+  -/
+  entities : Partial.Entities
+
+/--
+  Get the IDs of all policies which must be satisfied (for all possible
+  substitutions of the unknowns) and have the given `Effect`
+-/
+def Response.mustBeSatisfied (resp : Partial.Response) (eff : Effect) : Set PolicyID :=
+  Set.make (resp.residuals.filterMap (Residual.mustBeSatisfied eff))
+
+/--
+  Get the IDs of all policies which are, or may be, satisfied (for some
+  possible substitution of the unknowns) and have the given `Effect`
+-/
+def Response.mayBeSatisfied (resp : Partial.Response) (eff : Effect) : Set PolicyID :=
+  Set.make (resp.residuals.filterMap (Residual.mayBeSatisfied eff))
+
+/--
+  All `permit` policies which are definitely satisfied (for all possible
+  substitutions of the unknowns)
+-/
+def Response.knownPermits (resp : Partial.Response) : Set PolicyID :=
+  resp.mustBeSatisfied .permit
+
+/--
+  All `forbid` policies which are definitely satisfied (for all possible
+  substitutions of the unknowns)
+-/
+def Response.knownForbids (resp : Partial.Response) : Set PolicyID :=
+  resp.mustBeSatisfied .forbid
+
+/--
+  All `permit` policies which are, or may be, satisfied
+-/
+def Response.permits (resp : Partial.Response) : Set PolicyID :=
+  resp.mayBeSatisfied .permit
+
+/--
+  All `forbid` policies which are, or may be, satisfied
+-/
+def Response.forbids (resp : Partial.Response) : Set PolicyID :=
+  resp.mayBeSatisfied .forbid
+
+/--
+  All policies which definitely produce errors (for all possible substitutions
+  of the unknowns)
+-/
+def Response.errors (resp : Partial.Response) : List (PolicyID × Error) :=
+  resp.residuals.filterMap λ residual => match residual with
+    | .error id error => some (id, error)
+    | _ => none
+
+inductive Decision where
+  | allow -- definitely Allow, for any substitution of the unknowns
+  | deny -- definitely Deny, for any substitution of the unknowns
+  | unknown -- Allow and Deny are both possible, depending on substitution of the unknowns
+
+deriving instance Repr, DecidableEq for Decision
+
+def Response.decision (resp : Partial.Response) : Partial.Decision :=
+  if ¬ resp.knownForbids.isEmpty
+  then .deny -- there is a known forbid, we'll always get explicit deny
+  else if resp.permits.isEmpty
+  then .deny -- there are no permits that are even possibly satisfied
+  else if ¬ resp.forbids.isEmpty
+  then .unknown -- there's at least one forbid that may be satisfied, and at least one permit that may be (or is) satisfied
+  else if ¬ resp.knownPermits.isEmpty
+  then .allow -- there are no forbids that are even possibly satisfied, and at least one permit that is definitely satisfied
+  else .unknown -- there are no forbids that are even possibly satisfied, and at least one permit that may be satisfied
```

Comment:
It also took me a while to digest the logic here since there are so many conditionals and interchanging of empty vs. non-empty checks. Here's a potential rewrite that is more clear to me:
```suggestion
def Response.decision (resp : Partial.Response) : Partial.Decision :=
  if ¬ resp.knownForbids.isEmpty || resp.permits.isEmpty
  then .deny
  else if resp.forbids.isEmpty && ¬ resp.knownPermits.isEmpty
  then .allow
  else .unknown
```

Reply:
(Similar suggestions for `overapproximateDeterminingPolicies` and `underapproximateDeterminingPolicies` below.)

Reply:
I find the inline comments necessary to convince the reader (i.e. myself) of the correctness of this specification.  Your proposed rewrite is more concise but (to me) harder to convince myself that it is the correct spec.  Nonetheless if you don't find the existing version clear, that's a red flag that we're not meeting the goal of a clear (and clearly correct) spec.  Let's find a version that is clear to both of us.  How do you feel about the readability of the Rust equivalent of this code, at https://github.com/cedar-policy/cedar/blob/d016d23672a777c8d823a0383eeab68b61af7c49/cedar-policy-core/src/authorizer/partial_response.rs#L100?  Better or worse?

Reply:
I still find my rewrite more understandable -- but that could just be personal bias 🤷‍♀️ Looking at your code again, I think what was causing me confusion was maintaining the state of what is true/false (i.e. on the line after, the `if` the condition is true, but on all subsequent lines it is false). Possibly this could be fixed by adjusting indentation/line breaks.

I prefer the Rust version to the version here because, it makes it easier to keep track of what is true/false.

Reply:
More insight into my thought process: For each conditional, I would read the condition, the `then` branch, and the inline comment, and then think about it a little to see if I agreed. Then before moving to the `else` branch, I had to remind myself of the original condition(s), so I'd look 2 (and/or 4, 6, 8) lines up and mentally negate things. This became mentally tiring past two conditionals.

Reply:
FWIW,  I also find Kesha's rewrite clearer and easier to understand.

Reply:
Pushed another proposal, that looks a lot like Kesha's, but keeps the inline comments, and rearranges indentation and formatting to hopefully make things a bit clearer.  I also split the conditions that Kesha `||`'d together, because these are totally distinct cases in my mind and need totally separate comments, but if yall feel strongly I can put the `||` in.

Reply:
I also totally refactored the `overapproximate` and `underapproximate` methods to use `decision()` as a subroutine, and this makes them a lot clearer IMO

Reply:
@khieta and @emina let me know if the latest changes address your concerns

Reply:
Rewrite is much clearer to me. Although I'd still prefer to have a comment at the top rather than inline so it will render on https://cedar-policy.github.io/cedar-spec/docs/ 😉 

The refactoring in the `overapproximate` and `underapproximate` methods was a good move. They are much easier to understand now.

Reply:
Added some more doc comments (here and elsewhere throughout the file).  Lmk what you think